pkg/softtainter/softtainter.go

```diff
@@ -0,0 +1,532 @@
+
+	des := desv1.KubeDescheduler{}
+
+	descheduler.SetupPlugins()
+	policy, err := descheduler.LoadPolicyConfig(st.policyConfigFile, nil, pluginregistry.PluginRegistry)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	err = st.client.Get(ctx, client.ObjectKey{
+		Namespace: operatorclient.OperatorNamespace,
+		Name:      operatorclient.OperatorConfigName,
+	}, &des)
+	if err != nil {
+		logger.Error(err, "failed reading descheduler operator CR")
+		return reconcile.Result{}, err
+	}
+
+	if des.Spec.DeschedulingIntervalSeconds == nil {
+		return reconcile.Result{}, fmt.Errorf("descheduler should have an interval set")
+	}
+	st.resyncPeriod = time.Duration(*des.Spec.DeschedulingIntervalSeconds) * time.Second
+
+	var lnargs *nodeutilization.LowNodeUtilizationArgs
+
+	for _, p := range policy.Profiles {
+		if p.Name == string(desv1.RelieveAndMigrate) {
+			for _, pc := range p.PluginConfigs {
+				if pc.Name == nodeutilization.LowNodeUtilizationPluginName {
+					lnargs = pc.Args.(*nodeutilization.LowNodeUtilizationArgs)
+				}
+			}
+		}
+	}
+	if lnargs == nil {
+		err := fmt.Errorf("unable to read LowNodeUtilizationArgs")
+		logger.Error(err, "reconciliation failed")
+		return reconcile.Result{}, err
+	}
+
+	enableSoftTainter := false
+	if des.Spec.ProfileCustomizations != nil && des.Spec.ProfileCustomizations.DevEnableSoftTainter {
+		enableSoftTainter = true
+	}
+
+	st.args = &softTainterArgs{
+		useDeviationThresholds: lnargs.UseDeviationThresholds,
+		thresholds:             lnargs.Thresholds,
+		targetThresholds:       lnargs.TargetThresholds,
+		applySoftTaints:        enableSoftTainter,
+		mode:                   des.Spec.Mode,
+	}
+
+	if lnargs.MetricsUtilization == nil || lnargs.MetricsUtilization.Prometheus == nil {
+		err := fmt.Errorf("unable to read MetricsUtilization.Prometheus")
+		logger.Error(err, "reconciliation failed")
+		return reconcile.Result{}, err
+	}
+	st.promQuery = lnargs.MetricsUtilization.Prometheus.Query
+
+	var authToken *api.AuthToken
+	for _, provider := range policy.MetricsProviders {
+		if provider.Prometheus != nil {
+			st.promURL = provider.Prometheus.URL
+			authToken = provider.Prometheus.AuthToken
+		}
+	}
+	if st.promURL == "" {
+		err := fmt.Errorf("unable to read promURL")
+		logger.Error(err, "reconciliation failed")
+		return reconcile.Result{}, err
+	}
+
+	if authToken != nil {
+		err = st.reconcileSecretToken(ctx, authToken)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	} else {
+		err = st.reconcileInClusterSAToken()
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	st.resourceNames = getResourceNames(st.args.thresholds)
+
+	nodeSelector := labels.Everything()
+	if policy.NodeSelector != nil {
+		sel, err := labels.Parse(*policy.NodeSelector)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		nodeSelector = sel
+	}
+
+	nl := corev1.NodeList{}
+	err = st.client.List(ctx, &nl, &client.ListOptions{LabelSelector: nodeSelector})
```

Comment:
Is `List` getting listed from an informer cache or directly from a cluster? The former is required to reduce the traffic between the controller and the apiserver.

Reply:
Controller-runtime is client is a caching client (internally based on informers for registered kinds): all the operations are actually performed against a cache.
See `func getCacheOption` in `cmd/soft-tainter/main.go`.